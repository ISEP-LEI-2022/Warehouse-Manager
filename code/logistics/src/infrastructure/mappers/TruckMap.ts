import TruckDTO from "../../domain/dto/TruckDTO";
import { Truck } from "../../domain/aggregates";
import { expectedJSON } from "../../application/controllers/truck/ITruckController";

export default class TruckMap{
    public static toDTO(truck: Truck): TruckDTO {
        return {
            registration: truck.registration.value(),
            tare: truck.tare.value(),
            capacity: truck.capacity.value(),
            autonomy: truck.autonomy.value()
        } as TruckDTO;
    }

    public static toPersistence(truck: Truck): object{
        return {
            registration: truck.registration.value(),
            tare: truck.tare.value(),
            capacity: truck.capacity.value(),
            autonomy: truck.autonomy.value()
        };
    }

    public static toDomain(truckDTO: TruckDTO): Truck{
        return new Truck(
            truckDTO.autonomy,
            truckDTO.capacity,
            truckDTO.registration,
            truckDTO.tare,
        );
    }

    public static toJSON(truckDTO: TruckDTO): expectedJSON{
        return  {
            registration: truckDTO.registration,
            tare: truckDTO.tare,
            capacity: truckDTO.capacity,
            autonomy: truckDTO.autonomy
        };
    }

    public static toJSONArray(truckList: TruckDTO[]): expectedJSON[]{
        const truckListDTO : expectedJSON[] = [];
        for(const route of truckList){
            truckListDTO.push(TruckMap.toJSON(route));
        }
        return truckListDTO;
    }

}