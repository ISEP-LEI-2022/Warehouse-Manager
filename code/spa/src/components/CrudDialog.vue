<script setup lang="ts">
import { ref, computed } from "vue";

const emit = defineEmits(["submit", "newValue"]);

const open = () => {
  display.value = true;
};
const submit = () => {
  emit("submit", properties.value);
};

const disable_field = (field_name: string) => {
  if (field_name in props.disabled_fields) {
    return true;
  }
  return false;
};

function unCamelCase(str: string) {
  return (
    str
      // insert a space between lower & upper
      .replace(/([a-z])([A-Z])/g, "$1 $2")
      // space before last upper in a sequence followed by lower
      .replace(/\b([A-Z]+)([A-Z])([a-z])/, "$1 $2$3")
      // uppercase the first character
      .replace(/^./, function (str) {
        return str.toUpperCase();
      })
  );
}

const model_properties = computed(() => {
  const properties = [];
  if (props.model) {
    const keys = Object.getOwnPropertyNames(props.model);
    const values = Object.values(props.model);
    for (let i = 0; i < keys.length; i++) {
      properties.push({
        index: i,
        name: keys[i],
        value: values[i],
        type: typeof values[i],
      });
    }
  }
  return properties;
});

const props = defineProps<{
  title: string;
  edit: boolean;
  model: object;
  help_text_fields: {};
  disabled_fields: string[];
}>();

const properties = ref(model_properties);
const display = ref(false);
</script>

<template>
  <div style="margin-bottom: 1rem">
    <Dialog
      :header="props.title"
      v-model:visible="display"
      :breakpoints="{ '960px': '75vw' }"
      :style="{ width: '30vw' }"
      :modal="true"
    >
      <div class="card p-fluid">
        <div v-for="(property, index) in properties" :key="index" class="field">
          <strong
            ><label :for="property.name"
              >{{ unCamelCase(property.name) }} *</label
            ></strong
          >
          <InputText
            :id="property.name"
            :type="property.type"
            v-model="properties[index].value"
            :disabled="disable_field(property.name)"
          />
          <small v-if="props.help_text_fields[property.name]">{{
            props.help_text_fields[property.name]
          }}</small>
        </div>
      </div>
      <template #footer>
        <Button
          label="Save"
          @click="submit"
          icon="pi pi-check"
          class="p-button-outlined"
        />
      </template>
    </Dialog>
    <Button
      v-if="!edit"
      label="Add"
      icon="pi pi-external-link"
      style="width: auto"
      @click="open"
    />
    <Button v-else icon="pi pi-pencil" @click="open" />
  </div>
</template>

<style scoped lang="scss">
@import "@/assets/demo/styles/badges.scss";
</style>
