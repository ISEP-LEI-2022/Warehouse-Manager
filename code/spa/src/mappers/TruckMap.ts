import Truck from "@/models/truck";

export default class TruckMap {
  public static fromJSONArray(trucks: any): Truck[] {
    const truckList: Truck[] = [];
    for (const truck of trucks){
      truckList.push(new Truck(truck))
    }
    return truckList;
  }
}
