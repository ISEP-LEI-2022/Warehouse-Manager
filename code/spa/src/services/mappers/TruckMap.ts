import Truck from "@/models/truck";
import type TruckDTO from "../dtos/TruckDTO";

export default class TruckMap {
  public static fromDTOArray(trucks: Array<TruckDTO>): Truck[] {
    const truckList: Truck[] = [];
    if (trucks) {
      for (const truck of trucks) {
        truckList.push(
          new Truck(
            truck.registration,
            truck.capacity,
            truck.autonomy,
            truck.tare,
            truck.active
          )
        );
      }
    }
    return truckList;
  }

  public static fromDTO(truck: TruckDTO): Truck {
    const truckRes = new Truck(
      truck.registration,
      truck.capacity,
      truck.autonomy,
      truck.tare,
      truck.active
    );
    return truckRes;
  }

  public static fromAnyArray(truck: Array<any>): Truck {
    let obj = Object.assign({}, ...truck.map((x) => ({ [x.name]: x.value })));
    return new Truck(
      obj.Registration,
      obj.Capacity,
      obj.Autonomy,
      obj.Tare,
      obj.Active
    );
  }

  public static toJson(truck: Truck): string {
    return JSON.stringify(
      {
        registration: truck.Registration,
        capacity: Number(truck.Capacity),
        autonomy: Number(truck.Autonomy),
        tare: Number(truck.Tare),
        active: Boolean(truck.Active),
      },
      null,
      "\t"
    );
  }

  public static toDTO(truck: Truck): TruckDTO {
    return {
      registration: truck.Registration,
      capacity: Number(truck.Capacity),
      autonomy: Number(truck.Autonomy),
      tare: Number(truck.Tare),
      active: Boolean(truck.Active),
    };
  }

  public static empty(): Truck {
    return new Truck("XX-00-XX", 1, 1, 1, true);
  }
}
