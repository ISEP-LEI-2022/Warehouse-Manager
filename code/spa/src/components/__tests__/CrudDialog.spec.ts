import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import CrudDialog from '../CrudDialog.vue'
import TruckMap from '@/services/mappers/TruckMap'


describe('CrudDialog', () => {
  it('renders properly', () => {
    const wrapper = mount(CrudDialog,
      {
        props:
        {
          title: "Add Truck",
          edit: false,
          model: TruckMap.empty(),
          help_text_fields: {},
        }
    }
    )
    expect(wrapper.find('button').exists())
    expect(wrapper.find('#pv_id_27_header').exists())
    expect(wrapper.text()).toMatchSnapshot()
  })
})
