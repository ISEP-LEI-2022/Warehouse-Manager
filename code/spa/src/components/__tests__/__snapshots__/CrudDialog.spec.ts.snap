// Vitest Snapshot v1

exports[`CrudDialog > renders properly 1`] = `"Registration *Capacity *Autonomy *Tare *Active *"`;
